x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/IndexServiceAccountsTokenStore.java

```diff
@@ -130,22 +138,64 @@ public void createToken(Authentication authentication, CreateServiceAccountToken
-            .must(QueryBuilders.termQuery("username", accountId.asPrincipal()));
-        final SearchRequest request = client.prepareSearch(SECURITY_MAIN_ALIAS)
-            .setScroll(DEFAULT_KEEPALIVE_SETTING.get(getSettings()))
-            .setQuery(query)
-            .setSize(1000)
-            .setFetchSource(false)
-            .request();
-        request.indicesOptions().ignoreUnavailable();
+        final SecurityIndexManager frozenSecurityIndex = this.securityIndex.freeze();
+        if (false == frozenSecurityIndex.indexExists()) {
+            listener.onResponse(List.of());
+        } else if (false == frozenSecurityIndex.isAvailable()) {
+            listener.onFailure(frozenSecurityIndex.getUnavailableReason());
+        } else {
+            // TODO: wildcard support?
+            final BoolQueryBuilder query = QueryBuilders.boolQuery()
+                .filter(QueryBuilders.termQuery("doc_type", SERVICE_ACCOUNT_TOKEN_DOC_TYPE))
+                .must(QueryBuilders.termQuery("username", accountId.asPrincipal()));
+            final SearchRequest request = client.prepareSearch(SECURITY_MAIN_ALIAS)
+                .setScroll(DEFAULT_KEEPALIVE_SETTING.get(getSettings()))
+                .setQuery(query)
+                .setSize(1000)
+                .setFetchSource(false)
+                .request();
+            request.indicesOptions().ignoreUnavailable();
+
+            logger.trace("Searching tokens for service account [{}]", accountId);
+            ScrollHelper.fetchAllByEntity(client, request,
+                new ContextPreservingActionListener<>(client.threadPool().getThreadContext().newRestorableContext(false), listener),
+                hit -> extractTokenInfo(hit.getId(), accountId));
+        }
+    }
 
-        logger.trace("Searching tokens for service account [{}]", accountId);
-        ScrollHelper.fetchAllByEntity(client, request,
-            new ContextPreservingActionListener<>(client.threadPool().getThreadContext().newRestorableContext(false), listener),
-            hit -> extractTokenInfo(hit.getId(), accountId));
+    public void deleteToken(DeleteServiceAccountTokenRequest request, ActionListener<Boolean> listener) {
+        final SecurityIndexManager frozenSecurityIndex = this.securityIndex.freeze();
+        if (false == frozenSecurityIndex.indexExists()) {
+            listener.onResponse(false);
+        } else if (false == frozenSecurityIndex.isAvailable()) {
+            listener.onFailure(frozenSecurityIndex.getUnavailableReason());
+        } else {
+            final ServiceAccountId accountId = new ServiceAccountId(request.getNamespace(), request.getServiceName());
+            if (false == ServiceAccountService.isServiceAccountPrincipal(accountId.asPrincipal())) {
+                listener.onResponse(false);
+                return;
+            }
+            final ServiceAccountToken token = ServiceAccountToken.newToken(accountId, request.getTokenName());
```

Comment:
This feels weird, it's generating a new secret, but doesn't use it.

Reply:
It indeed is! It is a copy/paste code from "createToken", how surprise!

Reply:
Removed and just build the qualified token name which is all we need here.